scripts/object_conversion/seurat/seurat_scp_convenience.R

```diff
@@ -0,0 +1,343 @@
+#' are taken from "cell.embeddings" of appropriate dimensionality reduction slot (eg. t-SNE or UMAP).
+#' Cluster files can have 2 or 3 dimensions. This function cuts off coordinates after the 
+#' third dimension (eg. using only the first 3 PCs) if more are present. Can be used to create 
+#' cluster file for subset of all cells by passing vector of cell names to subset.cells.
+#' 
+#' Read more about cluster files at https://github.com/broadinstitute/single_cell_portal/wiki/Cluster-Files.
+#' 
+#' @param object Seurat object. 
+#' @param file.name File path to write (.tsv is appropriate). (Must be new name; will not overwrite existing file.)
+#' @param reduction.use Name of dimensional reduction to use for coordinates (default 'tsne').
+#' @param max.cols Number of coordinate columns/dimensions to include (only 2 or 3 are acceptable) (default 3 or 
+#'   2 if only 2 columns available).
+#' @param subset.cells Vector of cell names to use in subsetting cluster file (default NULL).
+#' 
+#' @examples
+#' \dontrun{
+#' generateClusterFile(seurat_obj, "/path/to/file.tsv", reduction.use = "umap")
+#' }
+generateClusterFile = function(object, file.name, reduction.use = 'tsne', 
+                               max.cols = 3, subset.cells = NULL) {
+  
+  if (file.exists(file.name)){
+    stop('Provided file path already exists. Please use different file.name.')
+  }
+  
+  # Get coordinates from cell embeddings 
+  s_version = packageVersion('Seurat')$major
+  if (s_version == 2) {
+    coords = GetCellEmbeddings(object, reduction.type = reduction.use)
+  } else if (s_version == 3) {
+    coords = Embeddings(object, reduction = reduction.use)
+  } else {
+    stop ('Only Seurat objects with major version >= 2 are supported.')
+  }
+  if (!(max.cols %in% c(2, 3))) {
+    stop("Only 2 or 3 coordinate columns allowed. Please specify different max.cols.")
+  }
+  
+  if (!is.null(subset.cells)) {
+    coords = coords[subset.cells,]
+  }
+  # Cut off extra columns
+  if (ncol(coords) > max.cols) {
+    coords = coords[,1:max.cols]
+    message(paste0("Keeping only first", max.cols, " dimensions of selected reduction."))
+  }
+  # Add cell names and rename columns
+  coords = cbind(row.names(coords), coords)
+  colnames(coords) = c('NAME', 'X', 'Y', 'Z')[1:ncol(coords)]
+  
+  # Insert data type row and reorder
+  data_types = c('TYPE', rep('numeric', ncol(coords) - 1))
+  coords = rbind(coords, data_types)
+  coords = coords[c(nrow(coords), 1:(nrow(coords) - 1)),]
+  
+  write.table(coords, file = file.name, sep = '\t', row.names = F, quote = F)
+  message(paste0("Finished writing ", file.name))
+}
+
+# Global variables for metadata related functions below
+# ***These may need to be updated periodically***
+required_meta_cols = c('biosample_id', 'donor_id', 'species', 'disease', 'organ', 
+                       'library_preparation_protocol', 'sex', 'biosample_type')
```

Comment:
Currently, 'species', 'disease', 'organ', 'library_preparation_protocol' are all required so their ontology_label columns are also required...
 
```suggestion 
required_meta_cols = c('biosample_id', 'donor_id', 'species', 'species__ontology_label', 'disease', 'disease__ontology_label', 'organ', 'organ__ontology_label', 
                       'library_preparation_protocol', 'library_preparation_protocol__ontology_label', 'sex', 'biosample_type')
```